sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -357,8 +368,33 @@ private[hive] object HadoopTableReader extends HiveInspectors with Logging {
         case oi: DoubleObjectInspector =>
           (value: Any, row: MutableRow, ordinal: Int) => row.setDouble(ordinal, oi.get(value))
         case oi: HiveVarcharObjectInspector =>
-          (value: Any, row: MutableRow, ordinal: Int) =>
-            row.setString(ordinal, oi.getPrimitiveJavaObject(value).getValue)
+          if (emptyStringsAsNulls) {
+            (value: Any, row: MutableRow, ordinal: Int) => {
+              val strValue = oi.getPrimitiveJavaObject(value).getValue
+              if (strValue.isEmpty) {
+                row.setString(ordinal, null)
+              } else {
+                row.setString(ordinal, strValue)
+              }
+            }
+          } else {
+            (value: Any, row: MutableRow, ordinal: Int) =>
+              row.setString(ordinal, oi.getPrimitiveJavaObject(value).getValue)
+          }
```

Comment:
You could also separate this into two `case`s, which may make the code maintenance with upstream changes a little easier.
```scala
case oi: HiveVarcharObjectInspector if emptyStringsAsNulls => ...
case oi: HiveVarcharObjectInspector =>
  (value: Any, row: MutableRow, ordinal: Int) =>
    row.setString(ordinal, oi.getPrimitiveJavaObject(value).getValue)
```